api/graphite.go

```diff
@@ -557,3 +460,95 @@ func (s *Server) metricsDeleteRemote(orgId int, query string, peer cluster.Node)
 
 	return resp.DeletedDefs, nil
 }
+
+// executePlan looks up the needed data, retrieves it, and then invokes the processing
+// note if you do something like sum(foo.*) and all of those metrics happen to be on another node,
+// we will collect all the indidividual series from the peer, and then sum here. that could be optimized
+func (s *Server) executePlan(orgId int, plan expr.Plan) ([]models.Series, error) {
+
+	type locatedDef struct {
+		def  idx.Archive
+		node cluster.Node
+	}
+	minFrom := uint32(math.MaxUint32)
+	var maxTo uint32
+	//locatedDefs[request][<def.id>]locatedDef
+	locatedDefs := make(map[expr.Req]map[string]locatedDef)
+
+	// note that different patterns to query can have different from / to, so they require different index lookups
+	// e.g. target=movingAvg(foo.*, "1h")&target=foo.*
+	// note that in this case we fetch foo.* twice. can be optimized later
+	for _, r := range plan.Reqs {
+		series, err := s.findSeries(orgId, []string{r.Query}, int64(r.From))
+		if err != nil {
+			return nil, err
+		}
+
+		locatedDefs[r] = make(map[string]locatedDef)
+
+		minFrom = util.Min(minFrom, r.From)
+		maxTo = util.Max(maxTo, r.To)
+
+		for _, s := range series {
+			for _, metric := range s.Series {
+				if !metric.Leaf {
+					continue
+				}
+				for _, def := range metric.Defs {
+					locatedDefs[r][def.Id] = locatedDef{def, s.Node}
+				}
+			}
+		}
+	}
+
+	var reqs []models.Req
+	for r, ldefs := range locatedDefs {
+		for _, locdef := range ldefs {
+			archive := locdef.def
+			// set consolidator that will be used to normalize raw data before feeding into processing functions
+			// not to be confused with runtime consolidation which happens after all processing.
+			fn := mdata.Aggregations.Get(archive.AggId).AggregationMethod[0]
+			consolidator := consolidation.Consolidator(fn) // we use the same number assignments so we can cast them
+			reqs = append(reqs, models.NewReq(
+				archive.Id, archive.Name, r.Query, r.From, r.To, plan.MaxDataPoints, uint32(archive.Interval), consolidator, locdef.node, archive.SchemaId, archive.AggId))
+		}
+	}
+	reqRenderSeriesCount.Value(len(reqs))
+	if len(reqs) == 0 {
```

Comment:
if only requests of which at least one series is found would get added to `locatedDefs` (as suggested in the above comment) then this check could be before the above loop and it would only need to test for `len(locatedDefs) == 0`